tests/avify4mtest.c

```diff
@@ -0,0 +1,196 @@
+// Copyright 2022 Google LLC. All rights reserved.
+// SPDX-License-Identifier: BSD-2-Clause
+
+#include "apps/shared/y4m.h"
+#include "avif/avif.h"
+
+#include <assert.h>
+#include <inttypes.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+//------------------------------------------------------------------------------
+
+// Returns true if image1 and image2 are identical.
+static avifBool compareYUVA(const avifImage * image1, const avifImage * image2)
+{
+    if (image1->width != image2->width || image1->height != image2->height || image1->depth != image2->depth ||
+        image1->yuvFormat != image2->yuvFormat || image1->yuvRange != image2->yuvRange) {
+        printf("ERROR: input mismatch\n");
+        return AVIF_FALSE;
+    } else if (image1->width * image1->height < 1) {
```

Comment:
Nit: Don't use `else` if the previous block ends with a return statement.

Also, I believe that `image1->width` and `image1->height` are both nonzero for a successfully decoded image. This property is true for non-grid images because in frame width and height are at least 1 in AV1. So we just need to check the code that decodes grid images to see if it disallows zero output image width or height.

Reply:
> Don't use `else` if the previous block ends with a return statement.

Sometimes I find it makes sense to put an `else` after a `return` from a code flow point of view (either a switch-like block series or very similar conditions).

> we just need to check the code that decodes grid images to see if it disallows zero output image width or height.

There are two checks in [`avifDecoderDataFillImageGrid()`](https://github.com/AOMediaCodec/libavif/blob/a8cd44119112b55f0433025d57a1ca245ec23d7c/src/read.c#L1285) that would fail if `outputWidth/Height` is `0` or if `(firstTile->image->width * grid->columns) < grid->outputWidth` so the test is not needed. An assertion is enough then.

Reply:
Omitting `else` after a `return` statement is a common recommendation. For example,

https://clang.llvm.org/extra/clang-tidy/checks/readability-else-after-return.html

I understand your argument, so I don't always pick this nit. In this particular case, we have a series of error checks. Therefore the early-exit recommendation is appropriate.

Thank you for the links to the two checks in `avifDecoderDataFillImageGrid()`. Just wanted to make sure I understand it correctly: the first check cannot fail if `outputWidth` or `outputHeight` is 0, right?

```
    if (((firstTile->image->width * grid->columns) < grid->outputWidth) ||
        ((firstTile->image->height * grid->rows) < grid->outputHeight)) {
        avifDiagnosticsPrintf(data->diag,
                              "Grid image tiles do not completely cover the image (HEIF (ISO/IEC 23008-12:2017), Section 6.6.2.3.1)");
        return AVIF_FALSE;
    }
```

On the other hand, I agree the second check will fail if `outputWidth` or `outputHeight` is 0:

```
    if (((firstTile->image->width * (grid->columns - 1)) >= grid->outputWidth) ||
        ((firstTile->image->height * (grid->rows - 1)) >= grid->outputHeight)) {
        avifDiagnosticsPrintf(data->diag,
                              "Grid image tiles in the rightmost column and bottommost row do not overlap the reconstructed image grid canvas. See MIAF (ISO/IEC 23000-22:2019), Section 7.3.11.4.2, Figure 2");
        return AVIF_FALSE;
    }
```

Note: I found the check I was referring to. It is the following code in the `avifParseImageGridBox()` function:

```
    if ((grid->outputWidth == 0) || (grid->outputHeight == 0)) {
        avifDiagnosticsPrintf(diag, "Grid box contains illegal dimensions: [%u x %u]", grid->outputWidth, grid->outputHeight);
        return AVIF_FALSE;
    }
```